### PR TITLE
virtual pets

### PR DESCRIPTION
**Pull request by Ben10Omintrix**
## About The Pull Request
players can now download their very own virtual pet through a new PDA app. this pet is called orbie and u can interact with him in alot of unique ways

https://github.com/tgstation/tgstation/assets/138636438/958042d4-08d8-4b39-9d2c-5f312cb2bae4

The app will require u to maintain ur pet's happiness and hunger.
interface

happiness can be increased by regularly grooming ur pet, feeding him, or u can arrange with other players playdates for ur pets, they can play with each other and both their happiness will increase 
plorb

u can get food from ur pet through ur pda, it will assign u a random drop zone location in the station u need to go to, after which u can obtain ur pet's food then ur PDA will spawn a virtual chocolate bar that ur pet loves to eat. it wont assign u dropzone locations that are restricted or hard to reach, however if the area it assigns u is a bit difficult to get to, u can reroll the location after a small cooldown

u can also level up ur pet to make it gain new helpful abilities and more cosmetic options. the main way to level up ur pet is by walking it, so u can have it follow u while u are doing ur job on the station and it will passively get exp. u get an increased exp modifier per step if ur pet is happy and not hungry. At level 2, ur pet will gain an ability to toggle lights and will also read outloud to u any PDA messages u recieve. at level 3, ur pet gains a camera ability. u can command it to take a photo afterwhich the picture will be saved directly in ur pda

u also have alot of customization options for ur pet!
custom
u can change its color, name, gender, and u can make it wear hats! u can unlock more hats for ur pet if u level it up further. these customizations change ur pet's hologram appearance as well as its profile picture on the pet network.

u can view how other player's virtual pets are progressing through the pet network. each time ur pet reaches a new milestone, an update will automatically be sent out on the network
update
if ur pet's milestones gets likes from other players, it will become happier

this app also allows u to program new tricks for ur pet. U can create a custom trick sequence, and change the trick's name. If u say the trick's name outloud to ur pet it will do the sequence u programmed. 
image

## Why It's Good For The Game
creates a new fun game on the pda players can play once every blue moon. 

## Changelog
:cl:
add: new virtual pet app on the pda
/:cl:

### CODE REVIEW

**Comment by lessthnthree** [comment]
finally the HoP has a friend to keep them company at the window

File: code/modules/mob/living/basic/pets/orbie/orbie_ai.dm [hunk 1]
```diff
@@ -0,0 +1,169 @@
+#define PET_PLAYTIME_COOLDOWN 2 MINUTES
+#define MESSAGE_EXPIRY_TIME 30 SECONDS
```
Comment (JohnFulpWillard) [inline code/modules/mob/living/basic/pets/orbie/orbie_ai.dm:2]:
```suggestion
#define PET_PLAYTIME_COOLDOWN (2 MINUTES)
#define MESSAGE_EXPIRY_TIME (30 SECONDS)
```

File: code/modules/modular_computers/file_system/programs/virtual_pet.dm [hunk 2]
```diff
@@ -0,0 +1,569 @@
...
+			break
+
+	data["pet_gender"] = ""
+	switch(pet.gender)
+		if(MALE)
+			data["pet_gender"] = "male"
+		if(FEMALE)
+			data["pet_gender"] = "female"
+		if(NEUTER)
+			data["pet_gender"] = "neutral"
```
Comment (JohnFulpWillard) [inline code/modules/modular_computers/file_system/programs/virtual_pet.dm:375]:
```suggestion
	data["pet_gender"] = pet.gender
```
You also don't take into account plural here

**Comment by Elli-Skala** [comment]
Finally death to PAIs

**Comment by zxaber** [comment]
I don't see anything in the code about how to obtain the app. I assume it's via the NT Downloader.

TBH, this seems more like it could be maint loot on a data disk; players would be able to copy the app from the disk and then share the disk with others. Might feel more special that way, I dunno.

**Comment by Elli-Skala** [comment]
> and will also read outloud to u any PDA messages u recieve. 
I can not wait intill I hear 5 of theses at once reading out a send all/coupon message

**Comment by DATA-xPUNGED** [comment]
> I don't see anything in the code about how to obtain the app. I assume it's via the NT Downloader.
> 
> TBH, this seems more like it could be maint loot on a data disk; players would be able to copy the app from the disk and then share the disk with others. Might feel more special that way, I dunno.

I agree with this, it's a really cool thing but it might be too much if a lot of people have it every round

**Review by Wallemations — APPROVED** [review]
yea sure the sprites suffice

**Comment by dwasint** [comment]
please stop making so many banger prs I can't take it anymore.

**Comment by deathrobotpunch** [comment]
ben100 is the true headcoder don’t believe the government’s lies

**Comment by Loafin34** [comment]
maint disks are where cool PDA apps go to die. dont put them there. also, any possibility you could reskin some pAIs to work as skins on orbie? that would be awesome. Nothing small, like the lizard, so that it's not hard to put hats on them.

**Comment by CMDR-Gungnir** [comment]
> maint disks are where cool PDA apps go to die. dont put them there. also, any possibility you could reskin some pAIs to work as skins on orbie? that would be awesome. Nothing small, like the lizard, so that it's not hard to put hats on them.

This. If it's a rare disk in maints you will never see it and it would be sad and a waste.

**Comment by DATA-xPUNGED** [comment]
oh hey here's an idea: black market uplink

**Comment by the-orange-cow** [comment]
Honestly, I don't see any reason to restrict something like this whatsoever.  For the people recommending it be locked behind something, what would be the value add of that?

**Comment by Exester509** [comment]
This opens the gates for an evil syndicate version in the future and I'm all for it

**Comment by Ben10Omintrix** [comment]
yes i have an evil version of the app coded for the syndi uplink but i thought it might be too much to put it all in this PR so if all goes well i might pr it in the future

**Comment by deathrobotpunch** [comment]
virtual pet named princess

**Comment by memecraft22** [comment]
this pda app actually looks really cool I hope we can see it in game soon

File: code/modules/modular_computers/file_system/programs/virtual_pet.dm (hunk 2 above)
Comment (Ben10Omintrix) [reply code/modules/modular_computers/file_system/programs/virtual_pet.dm:378]: the ui doesnt have enough space for 4 buttons so im just limiting it to these 3

**Comment by RaveRadbury** [comment]
pAI integration when?

**Comment by deathrobotpunch** [comment]
congratulations on the merge!

**Pull request merged**